### ISSUE TITLE
Maximum call stack exceeded with salePrice ref (video)

### ISSUE THREAD

**Issue by uturnr** [Code-Pop/vue-3-reactivity#2]
In the course video (but not the text below), the `salePrice` effect is declared after the `total` effect. This causes an infinite loop.

(Example of order in video)
```
effect(() => {
  total = salePrice.value * product.quantity
})
effect(() => {
  salePrice.value = product.price * 0.9
})
```

I added a check in the ref function to not set the value if it had not changed, which solved it for me:

```
    set value(newVal) {
      if (newVal !== raw) {
        raw = newVal
        trigger(r, 'value')
      }
    }
```

But then I came here to investigate and I noticed the difference, so I thought I'd point it out.

**Comment by Eugene930** [comment]
Same problem, and I find out that it's because the trigger function do not change the `activeEffect` variable when there is a nested trigger, [problem code here](https://github.com/Code-Pop/vue-3-reactivity/blob/b9721e40794a172f7162a3834eeb60e8900d20b6/06-ref.js#L31);

So I make some change in the `trigger` function:
```js
dep.forEach((innerEffect) => {
  effect(innerEffect);
});
``` 
this change the `activeEffect` variable when there is a deeper trigger

**Comment by MellowCobra** [comment]
Glad I'm not the only one who ran into this issue! Both of the above look like good solutions. I went ahead and put them into my own notes in case I come back to look at this later. Thanks for bringing this up!

**Comment by MellowCobra** [comment]
Following @Eugene930's example, should we also change the `effect` function? 
Currently in the tutorials, it sets `activeEffect` to null at the end of the function. This seems like if we add support for nested effect calls, this might be problematic as an inner `effect()` call will completely clear out the parent's effect.

Should it instead temporarily replace the active effect and then re-place it when it is done? Such as the following:
```javascript
function effect(eff) {
  const prevEffect = activeEffect
  activeEffect = eff
  activeEffect()
  activeEffect = prevEffect
}
```

**Comment by Eugene930** [comment]
> Following @Eugene930's example, should we also change the `effect` function?
> Currently in the tutorials, it sets `activeEffect` to null at the end of the function. This seems like if we add support for nested effect calls, this might be problematic as an inner `effect()` call will completely clear out the parent's effect.
> 
> Should it instead temporarily replace the active effect and then re-place it when it is done? Such as the following:
> 
> ```js
> function effect(eff) {
>   const prevEffect = activeEffect
>   activeEffect = eff
>   activeEffect()
>   activeEffect = prevEffect
> }
> ```

Yes we should change. I think you could take a look at [vue3's reactivity repo](https://github.com/vuejs/vue-next/tree/master/packages/reactivity), they use a better way: [stack](https://github.com/vuejs/vue-next/blob/fe5428db1207747886957b831d46d71ecb6fadaa/packages/reactivity/src/effect.ts#L99) , to achieve that.

**Comment by pinglu85** [comment]
I ran into the same issue. I tried @Eugene930's solution and it solved the problem for me ([codesandbox](https://codesandbox.io/s/brave-thunder-x5lbg)). However, it still causes an infinite loop when running in [codepen](https://codepen.io/pinglu/pen/mdOxMQo) and [jsbin](https://jsbin.com/kuqepun/edit?js,console). I can not figure out what causes this different behaviors. Any help would be appreciated.

**Comment by BorislavBorisov22** [comment]
I was just about to make a pr for a fix that I came up with, then I saw that there is already an issue about this. I simply introduced a new variable `skipTrack` and set it to true before trigger and then back to false once we have triggered the `dep` effects because I was not able to think of a case where we want to track any deps while triggering a setter.

```js
function reactivity() {
  const targetMap = new WeakMap();
  let activeEffect = null;
  let skipTrack = false;

  function effect(eff) {
    activeEffect = eff;
    activeEffect();
    activeEffect = null;
  }

  function track(target, key) {
    if (skipTrack || !activeEffect) {
      return;
    }

    let depsMap = targetMap.get(target);
    if (!depsMap) {
      targetMap.set(target, (depsMap = new Map()));
    }

    let dep = depsMap.get(key);
    if (!dep) {
      depsMap.set(key, (dep = new Set()));
    }

    dep.add(activeEffect);
  }

  function runEffects(dep) {
    skipTrack = true;
    dep.forEach(eff => eff());
    skipTrack = false;
  }

  function trigger(target, key) {
    const depsMap = targetMap.get(target);
    if (!depsMap) {
      return;
    }

    const dep = depsMap.get(key);
    if (!dep) {
      return;
    }

    runEffects(dep);
  }

  function reactive(target) {
    const handler = {
      get(target, key, receiver) {
        const result = Reflect.get(target, key, receiver);
        track(target, key);

        return result;
      },
      set(target, key, value, receiver) {
        const oldValue = target[key];
        const result = Reflect.set(target, key, value, receiver);
        if (result && oldValue !== value) {
          trigger(target, key);
        }

        return result;
      }
    }

    return new Proxy(target, handler);
  }

  function ref(raw) {
    const r = {
      get value() {
        track(r, 'value');
        return raw;
      },
      set value(newValue) {
        const oldValue = raw;
        raw = newValue;
        if (newValue !== oldValue) {
          trigger(r, 'value');
        }
      }
    }

    return r;
  }

  function computed(getter) {
    let result = ref(null);
    effect(() => {
      result.value = getter();
    });

    return result;
  }

  return { effect, reactive, ref, computed, };
}

const { effect, reactive, ref, computed } = reactivity();

const product = reactive({ price: 10, quantity: 2 });
const salePrice = ref(0);
let total = 0;

console.log(total, salePrice.value);
```
